subsys/bluetooth/audio/has_client.c

```diff
@@ -0,0 +1,626 @@
+	inst->control_point_subscription.ccc_handle = 0x0000;
+	inst->control_point_subscription.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	inst->control_point_subscription.disc_params = &inst->params.discover;
+	if (IS_ENABLED(CONFIG_BT_EATT) && properties & BT_GATT_CHRC_NOTIFY) {
+		inst->control_point_subscription.value = BT_GATT_CCC_INDICATE | BT_GATT_CCC_NOTIFY;
+	} else {
+		inst->control_point_subscription.value = BT_GATT_CCC_INDICATE;
+	}
+	atomic_set_bit(inst->control_point_subscription.flags, BT_GATT_SUBSCRIBE_FLAG_VOLATILE);
+
+	return bt_gatt_subscribe(inst->conn, &inst->control_point_subscription);
+}
+
+static uint8_t control_point_discover_cb(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+					 struct bt_gatt_discover_params *params)
+{
+	struct has_inst *inst = inst_by_conn(conn);
+	const struct bt_gatt_chrc *chrc;
+	int err;
+
+	__ASSERT(inst, "no instance for conn %p", (void *)conn);
+
+	BT_DBG("conn %p attr %p params %p", (void *)inst->conn, attr, params);
+
+	if (!attr) {
+		BT_INFO("Control Point not found");
+		discover_complete(inst);
+		return BT_GATT_ITER_STOP;
+	}
+
+	chrc = attr->user_data;
+
+	err = control_point_subscribe(inst, chrc->value_handle, chrc->properties);
+	if (err) {
+		BT_ERR("Subscribe failed (err %d)", err);
+		discover_failed(inst, err);
+	}
+
+	return BT_GATT_ITER_STOP;
+}
+
+static int control_point_discover(struct has_inst *inst)
+{
+	BT_DBG("conn %p", (void *)inst->conn);
+
+	(void)memset(&inst->params.discover, 0, sizeof(inst->params.discover));
+
+	(void)memcpy(&inst->params.uuid, BT_UUID_HAS_PRESET_CONTROL_POINT,
+		     sizeof(inst->params.uuid));
+	inst->params.discover.uuid = &inst->params.uuid.uuid;
+	inst->params.discover.func = control_point_discover_cb;
+	inst->params.discover.start_handle = BT_ATT_FIRST_ATTRIBUTE_HANDLE;
+	inst->params.discover.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	inst->params.discover.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+
+	return bt_gatt_discover(inst->conn, &inst->params.discover);
+}
+
+static void features_update(struct has_inst *inst, const void *data, uint16_t len)
+{
+	struct net_buf_simple buf;
+
+	net_buf_simple_init_with_data(&buf, (void *)data, len);
+
+	inst->has.features = net_buf_simple_pull_u8(&buf);
+
+	BT_DBG("conn %p features 0x%02x", (void *)inst->conn, inst->has.features);
+}
+
+static uint8_t features_read_cb(struct bt_conn *conn, uint8_t att_err,
+				struct bt_gatt_read_params *params, const void *data, uint16_t len)
+{
+	struct has_inst *inst = inst_by_conn(conn);
+	int err = att_err;
+
+	__ASSERT(inst, "no instance for conn %p", (void *)conn);
+
+	BT_DBG("conn %p att_err 0x%02x params %p data %p len %u", (void *)conn, att_err, params,
+	       data, len);
+
+	if (att_err != BT_ATT_ERR_SUCCESS || len == 0) {
+		goto fail;
+	}
+
+	features_update(inst, data, len);
+
+	if (!client_cb->active_index) {
+		/* Complete the discovery if client is not interested in preset updates */
```

Comment:
If I understand this correctly: If there is no callback for the active preset index, the control point characteristic will also not be discovered?
At first look, not discovering two characteristics based upon whether there is a callback for one of them, seems a bit drastic to me, but it may be a sensible way to do it.  But if so, this must be documented in the header file (for the active_index callback, possibly also for the discover function), that if the callback is not set, neither the control point nor the active present characteristic will be discovered.